### PR TITLE
Support IPv6 for nerdctl network

### LINKED ISSUE #1547: network create flag --ipv6 support
### What is the problem you're trying to solve

I am trying to implement containerd network with ipv6 traffic which is not supported currently is there any plan to address this?

### Describe the solution you'd like

if you support similar kind of feature like docker deamon or support nerdctl network create --ipv6 flag support it will help us resolve ipv6 traffic flow issue within contrainer.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yuchanns**
Fix #1547 

FYI: [cni-ipam](https://www.cni.dev/plugins/current/ipam/host-local/)

Tasks:
- [x] Support in the `network create` command.
- [x] Support `--ip6` argument for the `container run` command.
- [x] Integration tests.
- [x] Documentation.
- [ ] ...

Signed-off-by: Hanchin Hsieh me@yuchanns.xyz
Co-authored-by: Zheao Li <me@manjusaka.me>

### CODE REVIEW

**Comment by fahedouch** [comment]
this may inspire you https://github.com/containerd/nerdctl/pull/127/files :-D

**Comment by yuchanns** [comment]
@containerd/nerdctl-maintainers Hey guys I'm encountering issues. 

The CI cannot pass when allocating the IPv6 address but test cases work well locally on my laptop. Later @Zheaoli figured out that the problem is lacking IPv6 support in test runners. And I reproduced the scene on the local host after disabling IPv6 support and confirmed that.

The matter is that GHA seems not to support IPv6 on test runners yet (and not going to) according to actions/runner-images#668.

Any ideas?

File: cmd/nerdctl/container_run_network_linux_test.go [hunk 1]
```diff
@@ -575,3 +575,73 @@ func TestRunContainerWithMACAddress(t *testing.T) {
 		}
 	}
 }
+
+func TestRunContainerWithStaticIP6(t *testing.T) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/container_run_network_linux_test.go:579]: What about non-static?

Comment (AkihiroSuda) [inline cmd/nerdctl/container_run_network_linux_test.go:579]: IPv6 tests should require a new test CLI flag like `-test.ipv6=true`
https://github.com/containerd/nerdctl/blob/1642bef99f0931b79fe75fd54c1566e9e061b178/pkg/testutil/testutil.go#L507

**Comment by yuchanns** [comment]
Blocking by lacking IPv6 test support. There is proposal #2031

**Comment by AkihiroSuda** [comment]
> Blocked by lacking IPv6 test support. There is proposal #2031

I think we can merge this PR without CI as an experimental feature.

File: pkg/ocihook/ocihook.go [hunk 2]
```diff
@@ -221,6 +225,7 @@ type handlerOpts struct {
 	extraHosts        map[string]string // host:ip
 	containerIP       string
 	contianerMAC      string
```
Comment (aojea) [inline pkg/ocihook/ocihook.go:227]:
```suggestion
	containerMAC      string
```

**Comment by aojea** [comment]
> Blocked by lacking IPv6 test support. There is proposal #2031

the runners support IPv6 , you can run IPv6 internally in the machine, we have CI on kind running ipv6 only clusters.

What is not supported is IPv6 connectivity, the runners can not communicate externally with IPv6

**Comment by aojea** [comment]
can you rebase and push again the PR to take a look at the logs of the failures?

**Comment by yuchanns** [comment]
Fine. I will take a look later today.

File: cmd/nerdctl/container_run_network_linux_test.go (hunk 1 above)
Comment (yuchanns) [reply cmd/nerdctl/container_run_network_linux_test.go:489]: Make sense

**Comment by aojea** [comment]
it is a permissions problem 

```
ailed (add): failed to set bridge addr: could not add IP address
```

it looks like a permission problems , I can't remember now but I think that you may need to tune a bit more the way you run the tests inside the container

https://github.com/containerd/nerdctl/blob/ce742c94c68897f07f94c7d68c84d1b52b23fce2/.github/workflows/test.yml#L105

File: cmd/nerdctl/container_run_network_linux_test.go (hunk 1 above)
Comment (yuchanns) [reply cmd/nerdctl/container_run_network_linux_test.go:489]: It was my mistake to name it `static`. In the test table, an item contains an empty IP string that can be assigned a non-static IP6. 

https://github.com/containerd/nerdctl/blob/f29f5f538ca468397af5d0aeeb5d233e6ac98324/cmd/nerdctl/container_run_network_linux_test.go#L507-L511

Later, it is verified whether it is a valid IP6 address.

https://github.com/containerd/nerdctl/blob/f29f5f538ca468397af5d0aeeb5d233e6ac98324/cmd/nerdctl/container_run_network_linux_test.go#L545-L547

File: Dockerfile [hunk 3]
```diff
@@ -252,6 +252,7 @@ RUN perl -pi -e 's/multi-user.target/docker-entrypoint.target/g' /usr/local/lib/
   mkdir -p /etc/bash_completion.d && \
   nerdctl completion bash >/etc/bash_completion.d/nerdctl && \
   mkdir -p -m 0755 /etc/cni
+RUN sudo sysctl -w net.ipv6.conf.all.forwarding=1 && sudo sysctl -w net.ipv4.ip_forward=1
```
Comment (AkihiroSuda) [inline Dockerfile:255]: The applied configuration disappears after the `RUN` line, so you have to specify this via `docker run --sysctl`

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -104,6 +104,51 @@ jobs:
       - name: "Run integration tests"
         run: docker run -t --rm --privileged test-integration
 
+  test-integration-ipv6:
```
Comment (AkihiroSuda) [inline .github/workflows/test.yml:107]: We have too many tests now, can we just merge test-integration-ipv6 to test-integration?

Comment (AkihiroSuda) [reply .github/workflows/test.yml:107]: If not, tes-integration-ipv6 should be just executed for Ubuntu 22.04 and containerd v1.7

File: Dockerfile [hunk 5]
```diff
@@ -342,3 +342,7 @@ COPY ./Dockerfile.d/home_rootless_.config_systemd_user_containerd.service.d_port
 RUN chown -R rootless:rootless /home/rootless/.config
 
 FROM base AS demo
+
+FROM test-integration AS test-integration-ipv6
+CMD ["gotestsum", "--format=testname", "--rerun-fails=2", "--packages=github.com/containerd/nerdctl/cmd/nerdctl/...", \
+  "--", "-timeout=30m", "-args", "-test.kill-daemon", "-test.ipv6"]
```
Comment (AkihiroSuda) [inline Dockerfile:348]: The final stage must remain “demo”, as it is the default stage

File: cmd/nerdctl/compose_up_linux_test.go [hunk 6]
```diff
@@ -367,7 +367,7 @@ networks:
   default:
     ipam:
       config:
-        - subnet: 10.1.100.0/24
+        - subnet: 10.1.101.0/24
```
Comment (AkihiroSuda) [inline cmd/nerdctl/compose_up_linux_test.go:370]: ?

File: pkg/labels/labels.go [hunk 7]
```diff
@@ -98,6 +98,8 @@ const (
 	// Boolean value which can be parsed with strconv.ParseBool() is required.
 	// (like "nerdctl/default-network=true" or "nerdctl/default-network=false")
 	NerdctlDefaultNetwork = Prefix + "default-network"
+
+	IP6Address = Prefix + "ip6"
```
Comment (AkihiroSuda) [inline pkg/labels/labels.go:102]: This line should come after the “IPAddress” line

File: .github/workflows/test.yml (hunk 4 above)
Comment (yuchanns) [reply .github/workflows/test.yml:107]: Previously, it was merged. However, when using the `--network host` option in Docker to run integration tests, many non-IPv6 tests failed. Therefore, I separated it and ran it independently.

File: cmd/nerdctl/compose_up_linux_test.go (hunk 6 above)
Comment (yuchanns) [reply cmd/nerdctl/compose_up_linux_test.go:370]: This is unintentionally 😓 I'll fix it

File: pkg/labels/labels.go (hunk 7 above)
Comment (yuchanns) [reply pkg/labels/labels.go:102]: Make sense

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -200,6 +239,8 @@ jobs:
           sudo apt-get install -y expect
       - name: "Ensure that the integration test suite is compatible with Docker"
         run: go test -timeout 20m -v -exec sudo ./cmd/nerdctl/... -args -test.target=docker -test.kill-daemon
+      - name: "Ensure that the IPv6 integration test suite is compatible with Docker"
+        run: go test -timeout 20m -v -exec sudo ./cmd/nerdctl/... -args -test.target=docker -test.kill-daemon -test.ipv6
```
Comment (AkihiroSuda) [inline .github/workflows/test.yml:243]: Doesn't this need reconfiguring sysctl and `/etc/docker/daemon.json` ?

Comment (yuchanns) [reply .github/workflows/test.yml:243]: No. Fun fact 😂 
image

Comment (yuchanns) [reply .github/workflows/test.yml:243]: Somehow the integration-tests-ipv6 fails to run nested inside docker and qemu, despite sysctl and daemon being configured. Later @Zheaoli  figured out a solution by using `network=host`  to run it successfully. However, this leads to non-IPv6 test failure. So I have to separate the IPv6 test.

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -104,6 +104,45 @@ jobs:
...
+      - name: "Remove snap loopback devices (conflicts with our loopback devices in TestRunDevice)"
+        run: |
+          sudo systemctl disable --now snapd.service snapd.socket
+          sudo apt-get purge -y snapd
+          sudo losetup -Dv
+          sudo losetup -lv
+      - name: "Register QEMU (tonistiigi/binfmt)"
+        run: docker run --privileged --rm tonistiigi/binfmt --install all
+      - name: "Run integration tests"
+        run: docker run --network host -t --rm --privileged test-integration-ipv6
```
Comment (AkihiroSuda) [inline .github/workflows/test.yml:144]: Please add a comment to explain why we need (and why we can safely(?) use) `--network host`

File: .github/workflows/test.yml (hunk 8 above)
Comment (aojea) [reply .github/workflows/test.yml:243]: hmm, it should be able to run nested, but nested networks are complex and hard to debug, maybe you are missing certain sysctl or something, who knows

File: .github/workflows/test.yml (hunk 9 above)
Comment (yuchanns) [reply .github/workflows/test.yml:149]: Done.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Comment by yankay** [comment]
Thanks @yuchanns @AkihiroSuda
amazing

**Pull request merged**